README.md
# CCEAP: Covert Channel Educational Analysis Protocol

The *Covert Channel Educational Analysis Protocol* (CCEAP) is a simple network protocol designed for teaching covert channels (network steganography) to professionals and students. It is also an easy-to-use covert channel traffic generator.

The protocol is explicitly vulnerable against several [hiding patterns](http://ih-patterns.blogspot.de/p/introduction.html), i.e. patterns that represent hiding methods (steganographic methods that create covert channels, see [here](https://github.com/cdpxe/Network-Covert-Channels-A-University-level-Course/) for my GitHub online class on network steganography/covert channels). The protocol's structure is simple and self-explanatory and its implementation is kept at a minimum level of code lines to make it especially accessible to students. There is also a [documentation](https://github.com/cdpxe/CCEAP/tree/master/documentation) and an *academic publication* (published under the umbrella of ACM CCS 2016, download link is also in the documentation) available.

Please send requests and feedback to the author: Steffen Wendzel, [www.wendzel.de](http://www.wendzel.de) (wendzel (at) hs-worms (dot) de).

## Kickstart - a few Examples:

#### 1. A First Start
First, start the server, e.g. on your local host: `./server -P 9999`. This lets the server run on IP 127.0.0.1, port 9999.

Then, connect with the client: `./client -D 127.0.0.1 -P 9999`. This will make the client send ten simple standard messages to your server. The server should display the received packets' meta-data.

#### 2. Covert Timing Channel
Now, let us create a simple covert timing channel that we use to transfer the file `/etc/hosts`. And we want to encode 1 and 0 bits with 500ms and 1000ms. Therefore, we start the server again, and then run the client as follows using CCEAP's `iat_encode` tool which encodes files into inter-arrival times (this would represent the [Inter-arrival Time](http://ih-patterns.blogspot.de/p/blog-page_40.html) pattern):

`./client -D 127.0.0.1 -P 9999 -t ´./iat_encode /etc/hosts 500 1000´`

#### 3. Simple Covert Channel Traffic Generator
Of course, one could also use `dd` together with `/dev/random` as a source of randomness to create a file with random bits and use this file instead of `/etc/hosts`.
Similarly, the [Manipulated Message Ordering](http://ih-patterns.blogspot.de/p/p10-pdu-order-pattern.html) pattern can be represented using `./seq_encode /etc/hosts 256 2` instead of `./iat_encode 500 1000`. Check the [documentation](https://github.com/cdpxe/CCEAP/tree/master/documentation) for details.

More parameters of `client` can be obtained by running `./client -h`.

#### More Examples

Additional examples can be found in the [sample_exercises](https://github.com/cdpxe/CCEAP/tree/master/sample_exercises) directory and in the documentation.


CHANGELOG
CHANGELOG for the CCEAP tool

v. 0.6.2 : 2021-04-26
 * Improved welcome message and add -q for the client, too.

v. 0.6.1 : 2019-04-21 (Easter)
 * new parameter -T for provision of a default IAT value used for *all* packets
 * better error handling and error reporting
 * add random sequence number tool *rand_seq* for use with -s
 * do not send more packets than allowed by -c when packets are duplicated; thanks to J. Kammerlander for reporting this issue
 * increased maximum number of packets that can be sent
 * make the tool compile under Debian (use -lrt)
 * added `runserloop` target in Makefile to run the server in a loop

v. 0.6.0 : 2018-06-12
 * slightly improved client (help) output (thanks to student feedback)
 * fixed an error message in the client
 * increased number of packets that can be sent from 255 to 2048
 * add iat_encode tool to help encode messages as IAT values
 * allow the combination of -c with -s; loop sequence numbers if necessary and output all sequence numbers when -v is used
 * thanks to J. Kammerlander for reporting most of the issues for this release

v. 0.5.2 : 2017-03-30
 * added code to compile CCEAP under MacOS (not officially supported, just experimental)
 * performed additional small improvements within the documentation
 * fix header dependency in Makefile for C object files

v. 0.5.1 : 2017-02-03
 * performed additional improvements within documentation and Makefile
 * added three pages of sample exercises including solutions
 * added quite parameter -q for the server

v. 0.5.0 : 2016-10-18
 * added support for the Inter-arrival Time Pattern (and with that also an (indirect) support for the Rate/Throughput Pattern)
 * aligned port used in Makefile with port used in docs
 * improved formating issues for PDF documentation and introduced several detail updates within the documentation

v. 0.4.1 : 2016-07-26
 * Initial public release



main.h
/*
 * CCEAP: An Implementation of an Educational Protocol for Covert Channel
 *        Analysis Purposes
 *
 * The goal of this tool is to provide a simple, accessible implementation
 * for students. The tool demonstrates several network covert channel
 * vulnerabilities in a single communication protocol.
 *
 * Copyright (C) 2016-2021 Steffen Wendzel, steffen (at) wendzel (dot) de
 *                    http://www.wendzel.de
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Cf. `LICENSE' file.
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <err.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <pthread.h>
#include <errno.h>
#include <time.h>
#include <sys/time.h>
#include <ctype.h>
#include <netinet/tcp.h>
#include <inttypes.h>
#include <math.h>
#include <time.h>

#define CCEAP_VER	"0.6.2"
#define CCEAP_AUTHOR	"Steffen Wendzel, https://www.wendzel.de"
#define ERR_EXIT	2
#define OK_EXIT		0

#define USAGE_CLIENT	1
#define USAGE_SERVER	2

typedef struct {
	u_int8_t opt_identifier;		/* freely selectable ID of option */
	u_int8_t opt_type;			/* type of the optional header */
	u_int8_t opt_value;			/* value of the optional header */
	u_int8_t dummy;				/* dummy value to end up with 32 bytes */
} options_t;
#define MAX_NUM_OPTIONS 32
#define DEFAULT_IAT_VAL	1000000			/* 1sec */

typedef struct {
	#define MAX_NUM_PREDEF_SEQNOS 16384
	#define MAX_NUM_PREDEF_IATS 16384
	#define MAX_NUM_PACKETS 16384
	u_int8_t sequence_number;		/* random value to identify a connection (incremented) */
	u_int8_t number_of_options;		/* num. of optional header elements */
	u_int8_t destination_length;		/* length of destination value */
	u_int8_t dummy;				/* dummy value to end up with 32 bytes */
	#define DSTPADSIZ 8
	char destination_and_padding[DSTPADSIZ];/* variable length ASCII dest. ID, followed by zero padding so that
						 * the header ends up in a whole 32-bit word */
} cceap_header_t;

void *calloc_exit(int size, int elem);
void usage(int type);
void print_gpl(void);



server.c
/*
 * CCEAP: An Implementation of an Educational Protocol for Covert Channel
 *        Analysis Purposes
 *
 * The goal of this tool is to provide a simple, accessible implementation
 * for students. The tool demonstrates several network covert channel
 * vulnerabilities in a single communication protocol.
 *
 * Copyright (C) 2016-2019 Steffen Wendzel, steffen (at) wendzel (dot) de
 *                    https://www.wendzel.de
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Cf. `LICENSE' file.
 *
 */

#include "main.h"

#ifdef __MACH__
 #include <mach/clock.h>
 #include <mach/mach.h>
#endif

void print_time_diff(void)
{
	long ms;
	time_t s;
	struct timespec spec_now;
	static struct timespec spec_last;
	static int first_call = 1;
#ifdef __MACH__ /* code from Stackoverflow.com (Mac OS lacks clock_gettime()) */
	#warning "Including experimental code for MacOS. CCEAP runs best on Linux!"
	clock_serv_t cclock;
	mach_timespec_t mts;

	host_get_clock_service(mach_host_self(), CALENDAR_CLOCK, &cclock);
	clock_get_time(cclock, &mts);
	mach_port_deallocate(mach_task_self(), cclock);
	spec_now.tv_sec = mts.tv_sec;
	spec_now.tv_nsec = mts.tv_nsec;
#else
	clock_gettime(CLOCK_REALTIME, &spec_now);
#endif
	
	if (first_call) {
		printf("0.000\n");
		first_call = 0;
	} else {
		s  = spec_now.tv_sec - spec_last.tv_sec;
		ms = (spec_now.tv_nsec - spec_last.tv_nsec) / 1.0e6;	
		if (ms < 0) {
			ms = 1000 + ms;
			s -= 1;
		}
		printf("%"PRIdMAX".%03ld\n", (intmax_t)s, ms);
	}
	bcopy(&spec_now, &spec_last, sizeof(struct timespec));
}

void
inform_disconnected()
{
	fprintf(stderr, "The connection was closed by the foreign host.\n");
	fprintf(stderr, "Shutdown ..\n");
	exit(OK_EXIT);
}

int
main(int argc, char *argv[])
{
	int ch;
	int sockfd;
	int connfd;
	socklen_t salen = sizeof(struct sockaddr_in);
	int yup;
	struct sockaddr_in sa;
	cceap_header_t *hdr;
	int lst_port = 0;
	options_t *options_array;
	int verbose = 0;
	int quiet = 0;
	int max = 0;
	int ret = 0;
	int x;
	fd_set rset;
	char buf[1024];
	
	while ((ch = getopt(argc, argv, "vhP:q")) != -1) {
		switch (ch) {
		case 'v':
			verbose = 1;
			break;
		case 'P':
			/* TCP port to listen on */
			lst_port = atoi(optarg);
			break;
		case 'q':
			quiet = 1;
			break;
		case 'h':
		default:
			usage(USAGE_SERVER);
			/* NOTREACHED */
		}
	}

	if (!quiet)
		print_gpl();

	/* basic checks of provided parameters */
	if (lst_port >= 0xffff || lst_port < 1) {
		fprintf(stderr, "TCP listen port out of range or not specified.\n");
		exit(ERR_EXIT);
	}
	
	/* short welcome notice */
	if (quiet) {
		fprintf(stdout, "---\n");
	} else {
		fprintf(stdout, "CCEAP - Covert Channel Educational Analysis Protocol (Server)\n");
		fprintf(stdout, "   => version: " CCEAP_VER ", written by: " CCEAP_AUTHOR "\n");
	}
	
	/* summarize information if verbose mode is used */
	if (verbose) {
		printf("verbose output:\n");
		printf("* Using TCP port %d\n", lst_port);
	}
	
	if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0)
		err(ERR_EXIT, "socket()");

	if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &yup, sizeof(yup)) != 0)
		err(ERR_EXIT, "setsockopt(..., SO_REUSEADDR, ...)");
	
	/* prepare and create the connection to the peer */
	bzero(&sa, sizeof(struct sockaddr_in));
	sa.sin_family = AF_INET;
	sa.sin_port = htons(lst_port);
	sa.sin_addr.s_addr = INADDR_ANY;
	
	if (bind(sockfd, (struct sockaddr *) &sa, sizeof(struct sockaddr_in)) < 0)
		err(ERR_EXIT, "bind()"); /* socket for client connection */
	
	if (listen(sockfd, 1) < 0)
		err(ERR_EXIT, "listen()");
	
	if ((connfd = accept(sockfd, (struct sockaddr *) &sa, &salen)) < 0) {
		err(ERR_EXIT, "accept()");
	}
	
	max = connfd;
	for (;;) {
		FD_ZERO(&rset);
		FD_SET(connfd, &rset);
		
		ret = select(max + 1, &rset, NULL, NULL, NULL);
		if (ret) {
			int size_header = sizeof(cceap_header_t);
		
			bzero(buf, sizeof(buf));

			if (FD_ISSET(connfd, &rset)) {
				if ((x = recv(connfd, buf, sizeof(buf) - 1, 0)) == -1)
					err(ERR_EXIT, "recv()");
				if (x == 0)
					inform_disconnected();
				
				if (x >= size_header) {
					char dst_n_pad[DSTPADSIZ + 1] = { '\0' };
					
					hdr = (cceap_header_t *) buf;
					
					printf("received data (%d bytes):\n", x);
					printf(" > time diff to prev pkt: ");
						print_time_diff();
					printf(" > sequence number:       %d\n", hdr->sequence_number);
					printf(" > destination length:    %d\n", hdr->destination_length);
					printf(" > dummy value:           %d\n", hdr->dummy);
					strncpy(dst_n_pad, hdr->destination_and_padding, DSTPADSIZ);
					printf(" > destination + padding: %s\n", dst_n_pad);
					printf(" > number of options:     %d\n", hdr->number_of_options);
					if (hdr->number_of_options >= 1) {
						int i;
						
						/* check whether too many options */
						if ((x - sizeof(cceap_header_t)) > (MAX_NUM_OPTIONS * sizeof(options_t))) {
							fprintf(stderr, "received too many options\n");
							exit(ERR_EXIT);
						}
						/* check whether number_of_options is correct */
						if (hdr->number_of_options * sizeof(options_t) > (x - sizeof(cceap_header_t))) {
							fprintf(stderr, "did not receive enough bytes for %d options\n",
								hdr->number_of_options);
							exit(ERR_EXIT);
						} else {
							if (verbose) {
								printf("    > packet structure overview with optional space:\n"
								       "\t\t\t  %ld bytes required for handling the indicated %d options\n"
								       "\t\t\t  %ld bytes attached for options data, of which\n"
								       "\t\t\t  %ld bytes of received trailing data will be ignored\n",
									hdr->number_of_options * sizeof(options_t),
									hdr->number_of_options,
									x - sizeof(cceap_header_t),
									x - sizeof(cceap_header_t) - (hdr->number_of_options * sizeof(options_t))
									);
							}
						}
						
						options_array = (options_t *) (buf + sizeof(cceap_header_t));
						printf("    > options overview:\n");
						for (i = 0; i <  hdr->number_of_options; i++) {
							printf("\t\t\t  #%d: (identifier: %hhu, type: %hhu, value: %hhu)\n",
								i + 1,
								(options_array + i)->opt_identifier,
								(options_array + i)->opt_type,
								(options_array + i)->opt_value);
						}
					}
				} else {
					printf("did not implement caching of segments but received only part of a "
						"buffer -- sry ...\n");
					exit(ERR_EXIT);
				}
				
			} else {
				printf("huh?!\n");
				exit(ERR_EXIT);
			}
		} else if (ret == 0) {
			/* do nothing */
		} else { /* ret = -1 */
			if (errno == EINTR)
				continue;
			else
				err(ERR_EXIT, "select");
		}
	}
	
	close(sockfd);
	return OK_EXIT;
}




iat_encode.c
/* A dirty and undocumented hack to encode data as IAT values searated by commas.
 * License: GPLv3, see `LICENSE' file.
 *
 * This script can be used as a parameter for the CCEAP client.
 *
 * Use the following form:
 * $ ./client -t `./iat_encode [input file] [low time value (ms)] [high time value (ms)]`
 * e.g.:
 * $ ./client -t `./iat_encode message.txt 10 20`
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

#define PRINT_COMMA	putchar(',');
#define CHK_BIT(x)	if (x) { printf("%s", low); } else { printf("%s", high); }

int main(int argc, char *argv[])
{
	FILE *fp;
	char byte[1];
	char *low = NULL;
	char *high = NULL;
	int first = 1;
	extern char *__progname;

	if (argc <= 3) {
		fprintf(stderr, "usage: %s [input filename] [low_time (ms)] [high_time (ms)]\n", __progname);
		exit(1);
	}
	if ((fp = fopen(argv[1], "r")) == NULL) {
		perror("fopen");
		exit(1);
	}
	low = argv[2];
	high = argv[3];

	while (fread(byte, 1, 1, fp)) {
		if (first)
			first = 0;
		else
			PRINT_COMMA /* comma behind every new value */
		
		CHK_BIT (byte[0] & 0x01) PRINT_COMMA
		CHK_BIT (byte[0] & 0x02) PRINT_COMMA
		CHK_BIT (byte[0] & 0x04) PRINT_COMMA
		CHK_BIT (byte[0] & 0x08) PRINT_COMMA
		CHK_BIT (byte[0] & 0x10) PRINT_COMMA
		CHK_BIT (byte[0] & 0x20) PRINT_COMMA
		CHK_BIT (byte[0] & 0x40) PRINT_COMMA
		CHK_BIT (byte[0] & 0x80)
	}
	fclose(fp);

	return 0;
}



client.c
/*
 * CCEAP: An Implementation of an Educational Protocol for Covert Channel
 *        Analysis Purposes
 *
 * The goal of this tool is to provide a simple, accessible implementation
 * for students. The tool demonstrates several network covert channel
 * vulnerabilities in a single communication protocol.
 *
 * Copyright (C) 2016-2019 Steffen Wendzel, steffen (at) wendzel (dot) de
 *                    https://www.wendzel.de
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Cf. `LICENSE' file.
 *
 */

#include "main.h"

/* simply iterate through IAT array and use the next element to call usleep() */
void
wait_IAT_before_send(u_int32_t *IAT_array, int IAT_array_elements)
{
	static int next_element = 0;

	/*printf("sleeping %i [usec]\n", IAT_array[next_element]);*/
	usleep(IAT_array[next_element]);
	next_element++;
	if (next_element == IAT_array_elements)
		next_element = 0;
}

int
main(int argc, char *argv[])
{
	int ch;
	int sockfd;
	struct sockaddr_in sa;
	cceap_header_t hdr;
	cceap_header_t *pkt;
	int dst_port = 0;
	char *dst_addr = NULL;
	u_int8_t sequence_number = 0 /* default ISN is 0 */;
	u_int8_t exclude = 0, duplicate = 0; /* seq. no. to exclude/duplicate */
	u_int8_t dummy = 0;
	char destination[8];
	u_int8_t destination_length = 0;
	int ret_val;
	options_t options_array[MAX_NUM_OPTIONS];
	u_int8_t num_options = 0;
	int verbose = 0;
	int cnt;
	int num_of_pkts_to_send = 10;
	int sequence_number_set = 0, sequence_set = 0;
	u_int8_t sequence_number_array[MAX_NUM_PREDEF_SEQNOS] = { '\0' };
	int sequence_number_array_elements = 0;
	u_int32_t IAT_array[MAX_NUM_PREDEF_IATS] = { '\0' };
	int IAT_array_elements = 0;
	u_int32_t default_iat = DEFAULT_IAT_VAL;
	int IAT_set = 0;
	int arg_c_set = 0;
	int quiet = 0;
		
	while ((ch = getopt(argc, argv, "vhP:D:d:i:o:u:p:x:c:s:t:T:q")) != -1) {
		switch (ch) {
		case 'v':
			verbose = 1;
			break;
	/* options for TCP/IP */
		case 'P':
			/* TCP destination port */
			dst_port = atoi(optarg);
			break;
		case 'D':
			/* IP destination address */
			if (!(dst_addr = (char *) calloc_exit(strlen(optarg) + 1, sizeof(char))))
				err(1, "calloc");
			strncpy(dst_addr, optarg, strlen(optarg));
			break;
	/* options for CCEAP */
		case 'd':
			/* destination address for the header */
			if (strlen(optarg) > 8) {
				fprintf(stderr, "Destination address too long.\n");
				exit(ERR_EXIT);
			}
			strncpy(destination, optarg, strlen(optarg));
			destination_length = strlen(destination);
			break;
		case 'c':
			/* number of packets to send */
			if (atoi(optarg) >= 0 && atoi(optarg) < MAX_NUM_PACKETS) {
				num_of_pkts_to_send = atoi(optarg);
				arg_c_set = 1;
			} else {
				fprintf(stderr, "Number of packets to send is out of range (MAX=%i).\n",
					MAX_NUM_PACKETS);
				exit(ERR_EXIT);
			}
			break;
		case 'i':
			/* Initial Sequence Number */
			if (sequence_set) {
				fprintf(stderr, "-i cannot be combined with -s\n");
				exit(ERR_EXIT);
			}
			if (atoi(optarg) >= 0 && atoi(optarg) < 0xff) {
				sequence_number = atoi(optarg);
			} else {
				fprintf(stderr, "Initial sequence number out of range.\n");
				exit(ERR_EXIT);
			}
			sequence_number_set = 1;
			break;
		case 's':
			/* define sequence of sequence_numbers for Message Order Manipulation
			 * in the form: id_1,id_2,id_3,... */
			if (sequence_number_set) {
				fprintf(stderr, "-s cannot be combined with -i\n");
				exit(ERR_EXIT);
			}
			
			{
				/* sequence_number_array[MAX_NUM_PREDEF_.....];
				sequence_number_array_elements = 0; */
				char *token;
				char *ptr_optarg;
				char *saveptr;
				
				bzero(sequence_number_array, sizeof(sequence_number_array));
				
				for (ptr_optarg = optarg; ; ptr_optarg = NULL) {
					int digit;
					
					token = strtok_r(ptr_optarg, ",", &saveptr);
					if (token == NULL)
						break;
					
					digit = token[0];
					if (isdigit(digit) == 0) {
						fprintf(stderr, "in parameter -s: '%c' is not a digit\n",
							token[0]);
						exit(ERR_EXIT);
					}
					
					sequence_number_array[sequence_number_array_elements] = atoi(token);
					sequence_number_array_elements++;
					if (sequence_number_array_elements > MAX_NUM_PREDEF_SEQNOS) {
						fprintf(stderr, "Too many sequence numbers "
							"(%i but max is %i) used in -s\n",
							sequence_number_array_elements,
							MAX_NUM_PREDEF_SEQNOS);
						exit(ERR_EXIT);
					} /*else {
						fprintf(stderr, "sequence_number_array_elements=%d <= %d\n",
							sequence_number_array_elements, MAX_NUM_PREDEF_SEQNOS);
					}*/
				}
			}
			if (sequence_number_array_elements > 0) {
				sequence_set = 1;
				if (!arg_c_set) {
					num_of_pkts_to_send = sequence_number_array_elements;
				}
			}
			break;
		case 't':
			/* define IATs */
			if (IAT_set || default_iat != DEFAULT_IAT_VAL) {
				fprintf(stderr, "IAT values were already set by previous parameter(s).\n");
				exit(ERR_EXIT);
			}
			{
				char *token;
				char *ptr_optarg;
				char *saveptr;
				
				bzero(IAT_array, sizeof(IAT_array));
				
				for (ptr_optarg = optarg; ; ptr_optarg = NULL) {
					int digit;
					
					token = strtok_r(ptr_optarg, ",", &saveptr);
					if (token == NULL)
						break;
					
					digit = token[0];
					if (isdigit(digit) == 0) {
						fprintf(stderr, "in parameter -t: '%c' is not a digit\n",
							token[0]);
						exit(ERR_EXIT);
					}
					
					IAT_array[IAT_array_elements] = atoi(token)*1000;
					IAT_array_elements++;
					if (IAT_array_elements > MAX_NUM_PREDEF_IATS) {
						fprintf(stderr, "Too many IAT values given in -t\n");
						exit(ERR_EXIT);
					} /*else {
						fprintf(stderr, "IAT_array_elements=%d <= %d\n",
							IAT_array_elements, MAX_NUM_PREDEF_IATS);
					}*/
				}
			}
			if (IAT_array_elements > 0) {
				IAT_set = 1;
				if (!arg_c_set) {
					num_of_pkts_to_send = IAT_array_elements;
				}
			}
			break;
		case 'T':
			if (IAT_set || default_iat != DEFAULT_IAT_VAL) {
				fprintf(stderr, "IAT values were already set by previous parameter(s).\n");
				exit(ERR_EXIT);
			}
			
			if (atoi(optarg) >= 0 && atoi(optarg) <= 10000) {
				default_iat = atoi(optarg)*1000;
			} else {
				fprintf(stderr, "Value for -T must be within 1 and 10000.\n");
				exit(ERR_EXIT);
			}
			break;
		case 'p':
			/* duplicate packet with sequence_number provided via optarg */
			if (atoi(optarg) >= 0 && atoi(optarg) < 0xff) {
				duplicate = atoi(optarg);
			} else {
				fprintf(stderr, "Duplicate sequence_number out of range.\n");
				exit(ERR_EXIT);
			}
			break;
		case 'x':
			/* drop packet (do not send it) with provided sequence_number */
			if (atoi(optarg) >= 0 && atoi(optarg) < 0xff) {
				exclude = atoi(optarg);
			} else {
				fprintf(stderr, "Exclusion sequence_number out of range.\n");
				exit(ERR_EXIT);
			}
			break;
		case 'u':
			/* unusued `dummy' field in main header */
			if (atoi(optarg) >= 0 && atoi(optarg) < 0xff) {
				dummy = atoi(optarg);
			} else {
				fprintf(stderr, "Dummy field value out of range.\n");
				exit(ERR_EXIT);
			}
			break;
		case 'o':
			/* optional header components in the form
			 * id_1,ty_1,val_1:id_2,ty_2,val_2:id_3,ty_3,val_3,...
			 */
			{
				char *token_1, *token_2;
				char *ptr_optarg_1, *ptr_optarg_2;
				char *saveptr_1, *saveptr_2;
				
				bzero(options_array, sizeof(options_array));
				
				for (ptr_optarg_1 = optarg; ; ptr_optarg_1 = NULL) {
					int num_elements = 0;
					
					token_1 = strtok_r(ptr_optarg_1, "/", &saveptr_1);
					if (token_1 == NULL)
						break;
					//printf("token_1: %s\n", token_1);
					
					for (num_elements = 0, ptr_optarg_2 = token_1; ; ptr_optarg_2 = NULL) {
						int digit;
						
						if (num_elements == MAX_NUM_OPTIONS) {
							fprintf(stderr, "too many options (max: %d)\n",
								MAX_NUM_OPTIONS);
							exit(ERR_EXIT);
						}
						
						token_2 = strtok_r(ptr_optarg_2, ",", &saveptr_2);
						if (token_2 == NULL)
							break;
						//printf("\ttoken_2: %s\n", token_2);
						digit = token_2[0];
						if (isdigit(digit) == 0) {
							fprintf(stderr, "in parameter -o: '%c' is not a digit\n",
								token_2[0]);
							exit(ERR_EXIT);
						}
						switch (num_elements) {
						case 0: /* option's identifier */
							options_array[num_options].opt_identifier = atoi(token_2);
							break;
						case 1: /* option's type */
							options_array[num_options].opt_type = atoi(token_2);
							break;
						case 2: /* option's value */
							options_array[num_options].opt_value = atoi(token_2);
							num_options++;
							break;
						default:
							fprintf(stderr, "more than three parameters not permitted"
									" per option for parameter -o\n");
							exit(ERR_EXIT);
							break;
						}
						num_elements++;
					}
					if (num_elements < 3) {
						fprintf(stderr, "three parameters are required per option for "
								"parameter -o.\n");
						exit(ERR_EXIT);
					}
				
				}
			}
			break;
		case 'q':
			quiet = 1;
			break;
		case 'h':
		default:
			usage(USAGE_CLIENT);
			/* NOTREACHED */
		}
	}
	
	if (!quiet)
		print_gpl();
	else {
		fprintf(stdout, "CCEAP - Covert Channel Educational Analysis Protocol (Server)\n");
		fprintf(stdout, "   => version: " CCEAP_VER ", written by: " CCEAP_AUTHOR "\n");
	}
	
	/* basic checks of provided parameters */
	if (dst_port >= 0xffff || dst_port < 1) {
		fprintf(stderr, "destination port out of range or not specified.\n");
		exit(ERR_EXIT);
	}
	
	if (dst_addr == NULL) {
		fprintf(stderr, "No destination address specified.\n");
		exit(ERR_EXIT);
	}
	
	/* if no ISN was specified and also no list of seq.no. was defined, then
	 * start with ISN=0 (incremental mode) */
	if (sequence_set == 0 /* no pre-defined sequence was provided */
	 && sequence_number_set == 0 /* no ISN was provided either */) {
	 	if (verbose) {
	 		printf("setting ISN to 1 and will use incremental sequence numbers.\n");
	 	}
		sequence_number = 1;
		sequence_number_set = 1; /* we pretent as if the user provided -i 1 */
	}

	/* summarize information if verbose mode is used */
	if (verbose) {
		/* short welcome notice */
		printf("CCEAP - Covert Channel Educational Analysis Protocol (Client)\n");
		printf("   => version: " CCEAP_VER ", written by: " CCEAP_AUTHOR "\n");
		putchar('\n');
		printf("* Sending %d packets\n"
		       "* Dummy value: %d\n"
		       "* Initial seq. no. to start with (relevant only if *not* in sequence mode): %d\n"
		       "* Seq. no. to duplicate / exclude: %d / %d\n",
		       	num_of_pkts_to_send, dummy, sequence_number,
		       	duplicate, exclude);
		if (sequence_set) {
			int q;
			
			printf("* Sequence of sequence_numbers to use: ");
			for (q = 0; q < num_of_pkts_to_send; q++) {
				printf("%d%c",
				 sequence_number_array[q % sequence_number_array_elements],
				 (((q + 1) % sequence_number_array_elements) ? ',' : '/')
				);
			}
			putchar('\n');
		}
		printf("* Determined %d options...", num_options);
		if (num_options > 0) {
			int i;
			putchar(' ');
			for (i = 0; i < num_options; i++) {
				printf("(%d,%d,%d),",
					options_array[i].opt_identifier,
					options_array[i].opt_type,
					options_array[i].opt_value);
			}
		}
		putchar('\n');
	}

	/* prepare and create the connection to the peer */
	bzero(&sa, sizeof(struct sockaddr_in));
	sa.sin_family = AF_INET;
	sa.sin_port = htons(dst_port);
	ret_val = inet_pton(AF_INET, dst_addr, &sa.sin_addr);
	if (!ret_val) {
		if (ret_val == 0) {
			fprintf(stderr, "Desination address not a valid character string.\n");
			exit(ERR_EXIT);
		} else {
			fprintf(stderr, "Invalid address family for inet_pton().\n");
			exit(ERR_EXIT);
		}
	}
	
	/* socket for client connection */
	if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0)
		err(ERR_EXIT, "socket");
	
	printf("connecting ... ");
	if (connect(sockfd, (struct sockaddr *) &sa, sizeof(struct sockaddr_in)) != 0)
		err(ERR_EXIT, "connect");
	printf("connected.\n");
	
	/* create protocol header */
	bzero(&hdr, sizeof(cceap_header_t));
	if (sequence_number_set)
		hdr.sequence_number = sequence_number;
	else /* sequence pre-defined */
		hdr.sequence_number = sequence_number_array[0];
	
	/* place padding bytes ('X'es) */
	strncpy(hdr.destination_and_padding, "XXXXXXXX", 8);
	/* now override 1-8 bytes of the padding (so that padding remains, if necessasry) */
	if (destination_length > 0) {
		strncpy(hdr.destination_and_padding, destination, destination_length);
		hdr.destination_length = destination_length;
	}
	
	/* overwrite dummy field value if necessary */
	if (dummy != 0) {
		hdr.dummy = dummy;
	}
	
	/* add number of options to MAIN header */
	hdr.number_of_options = num_options;
	
	/* make one chunk of data out of HEADER + OPTIONS */
	pkt = (cceap_header_t *) calloc_exit(1, sizeof(cceap_header_t) + (sizeof(options_t) * num_options));
	memcpy(pkt, &hdr, sizeof(hdr));
	memcpy(pkt + 1, &options_array, sizeof(options_t)*num_options);
	
	/* send main header + options in one chunk to destination */
	printf("Sending: ");

	for (cnt = 0; cnt < num_of_pkts_to_send; cnt++) {
	
		if (exclude && pkt->sequence_number == exclude) {
			printf("(Excluded packet)"); fflush(stdout);
			if (IAT_set)
				wait_IAT_before_send(IAT_array, IAT_array_elements);
			else
				usleep(default_iat);
		} else {
			if (send(sockfd, pkt, sizeof(cceap_header_t) + (sizeof(options_t) * num_options), 0) < 0)
				err(ERR_EXIT, "send");
			putchar('.'); fflush(stdout);
			if (IAT_set)
				wait_IAT_before_send(IAT_array, IAT_array_elements);
			else
				usleep(default_iat);
		}
		
		if (duplicate && pkt->sequence_number == duplicate) {
			/* only duplicate if there is a packet left that we are allowed to send */
			if ((cnt+1) < num_of_pkts_to_send) {
				cnt++;
				if (send(sockfd, pkt, sizeof(cceap_header_t) + (sizeof(options_t) * num_options), 0) < 0)
					err(ERR_EXIT, "send");
				printf("(Duplicated packet)"); fflush(stdout);
				if (IAT_set)
					wait_IAT_before_send(IAT_array, IAT_array_elements);
				else
					usleep(default_iat);
			} else {
				printf("(Prevented packet duplication due to limit (%i packets)\n", num_of_pkts_to_send);
				fflush(stdout);
			}
		}
		
		if (sequence_number_set) {
			pkt->sequence_number++;
			/* reset sequence_number in case of overrun */
			if (pkt->sequence_number == 0xff) {
				pkt->sequence_number = 0;
			}
		} else { /* sequence mode */
			pkt->sequence_number = sequence_number_array[(cnt+1) % sequence_number_array_elements];
		}
	}
	printf("\nsending done.\n");
	
	close(sockfd);
	return OK_EXIT;
}


